config-controller/pkg/client/client.go

```diff
@@ -0,0 +1,303 @@
+		err := retry.WithRetry(func() error {
+			var innerErr error
+			gc, innerErr = newGrpcClient(ctx)
+			if innerErr != nil {
+				getLogger(ctx).Error(innerErr, "Failed to connect to Central")
+			}
+
+			c.svc = gc
+
+			if innerErr = c.EnsureFresh(ctx); innerErr != nil {
+				getLogger(ctx).Error(innerErr, "Failed to initialize client")
+			}
+
+			return innerErr
+		}, retry.Tries(10), retry.WithExponentialBackoff())
+
+		if err != nil {
+			return nil, errors.Wrap(err, "could not initialize policy client")
+		}
+	} else {
+		if err := c.EnsureFresh(ctx); err != nil {
+			getLogger(ctx).Error(err, "Failed to initialize client")
+		}
+	}
+
+	return &c, nil
+}
+
+func (c *client) ListPolicies(ctx context.Context) ([]*storage.Policy, error) {
+	policies := make([]*storage.Policy, 0, len(c.cache))
+	for _, value := range c.cache {
+		policies = append(policies, value)
+	}
+	return policies, nil
+}
+
+func (c *client) GetPolicy(ctx context.Context, name string) (*storage.Policy, bool, error) {
+	policy, exists := c.cache[name]
+	return policy, exists, nil
+}
+
+func (c *client) CreatePolicy(ctx context.Context, policy *storage.Policy) (*storage.Policy, error) {
+	getLogger(ctx).Info("POST", "policyName", policy.Name)
+	createdPolicy, err := c.svc.PostPolicy(ctx, policy)
+
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("Failed to POST policy '%s'", policy.Name))
+	}
+
+	c.cache[createdPolicy.Name] = createdPolicy
+
+	return createdPolicy, nil
+}
+
+func (c *client) UpdatePolicy(ctx context.Context, policy *storage.Policy) error {
+	getLogger(ctx).Info("PUT", "policyName", policy.Name)
+	err := c.svc.PutPolicy(ctx, policy)
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to PUT policy")
+	}
+
+	c.cache[policy.Name] = policy
+
+	return nil
+}
+
+func (c *client) FlushCache(ctx context.Context) error {
+	if time.Since(c.lastUpdated).Seconds() < 10 {
+		// Don't flush the cache more often than every 10s
+		return nil
+	}
+
+	getLogger(ctx).Info("Flushing policy cache")
+
+	getLogger(ctx).Info("LIST")
+	allPolicies, err := c.svc.ListPolicies(ctx)
+	if err != nil {
+		return errors.Wrap(err, "Failed to list policies")
+	}
+
+	newCache := make(map[string]*storage.Policy, len(allPolicies))
+
+	for _, listPolicy := range allPolicies {
+		getLogger(ctx).Info("GET", "ID", listPolicy.Id)
+		policy, err := c.svc.GetPolicy(ctx, listPolicy.Id)
+		if err != nil {
+			return errors.Wrapf(err, "Failed to fetch policy %s", listPolicy.Id)
+		}
+		newCache[policy.Name] = policy
+	}
+
+	c.cache = newCache
+	c.lastUpdated = time.Now()
+
+	return nil
+}
+
+func (c *client) EnsureFresh(ctx context.Context) error {
+	// Make sure token isn't expired before flushing cache
+	if err := c.svc.TokenExchange(ctx); err != nil {
+		return err
+	}
+
+	if time.Since(c.lastUpdated).Minutes() > 5.0 {
+		return c.FlushCache(ctx)
+	}
+
+	return nil
+}
+
+func getLogger(ctx context.Context) logr.Logger {
+	return log.FromContext(ctx).WithName("central-client")
```

Comment:
Naming is hard, but better name? Config-controller-client?

Reply:
My thinking is that this is the logger for the module that represents the client to Central.  Can you perhaps describe why `central-client` doesn't sound right to you?